src/server/pps/server/pipeline_controller.go

```diff
@@ -242,95 +356,98 @@ func (pc *pipelineController) fetchState() (*pps.PipelineInfo, context.Context,
 	return pi, pachClient.Ctx(), nil
 }
 
-func (pc *pipelineController) transitionStates(ctx context.Context, pi *pps.PipelineInfo, rc *v1.ReplicationController) error {
-	// Bring 'pipeline' into the correct state by taking appropriate action
+func evaluate(pi *pps.PipelineInfo, rc *v1.ReplicationController) (pps.PipelineState, []sideEffect, error) {
 	if pi.State == pps.PipelineState_PIPELINE_FAILURE {
-		// pipeline fails if it encounters an unrecoverable error
-		if err := pc.finishPipelineOutputCommits(ctx, pi); err != nil {
-			return err
-		}
-		// deletePipelineResources calls cancelMonitor() and cancelCrashingMonitor()
-		// in addition to deleting the RC, so those calls aren't necessary here.
-		if err := pc.deletePipelineResources(); err != nil {
-			// retry, but the pipeline has already failed
-			return stepError{
-				error: errors.Wrap(err, "error deleting resources for failing pipeline"),
-				retry: true,
-			}
-		}
-		return nil
+		return pps.PipelineState_PIPELINE_FAILURE,
+			[]sideEffect{
+				FinishCommitsSideEffect(),
+				ResourcesSideEffect(sideEffectToggle_DOWN),
+			}, nil
 	}
 	if pi.State == pps.PipelineState_PIPELINE_STARTING || pi.State == pps.PipelineState_PIPELINE_RESTARTING {
+		sideEffects := make([]sideEffect, 0)
 		if rc == nil {
-			// default: old RC (if any) is down but new RC is not up yet
-			if err := pc.kd.CreatePipelineResources(ctx, pi); err != nil {
-				return err
-			}
+			sideEffects = append(sideEffects, ResourcesSideEffect(sideEffectToggle_UP))
 		}
 		if pi.Stopped {
-			return pc.setPipelineState(ctx, pi.SpecCommit, pps.PipelineState_PIPELINE_PAUSED, "")
+			return pps.PipelineState_PIPELINE_PAUSED, sideEffects, nil
 		}
-		pc.stopCrashingPipelineMonitor()
-		// trigger another event
-		target := pps.PipelineState_PIPELINE_RUNNING
+		sideEffects = append(sideEffects, CrashingMonitorSideEffect(sideEffectToggle_DOWN))
 		if pi.Details.Autoscaling && pi.State == pps.PipelineState_PIPELINE_STARTING {
-			// start in standby
-			target = pps.PipelineState_PIPELINE_STANDBY
+			return pps.PipelineState_PIPELINE_STANDBY, sideEffects, nil
 		}
-		return pc.setPipelineState(ctx, pi.SpecCommit, target, "")
+		return pps.PipelineState_PIPELINE_RUNNING, sideEffects, nil
 	}
 	if rc == nil {
 		// may happen if an external system deletes the RC
-		return pc.restartPipeline(ctx, pi, rc, "missing RC")
+		return pi.State, []sideEffect{RestartSideEffect()}, nil
```

Comment:
probably doesn't make sense to change in this PR, but the restart code also changes the pipeline's state. To better adhere to this model, I think it could make sense to return RESTARTING here and move that out of the actual restart code. Right now, if the restart side effect code modified the actual `pipelineInfo` passed in rather than just the database, this would be a bug

Reply:
right. And more generally the side effects shouldn't be updating the pipelineInfo in the DB